### PR TITLE
Transaction reduce relay feature.

### PR DESCRIPTION
**Pull request by gregtatcam**
## High Level Overview of Change

A node randomly selects peers to relay a transaction to.
It aggregates and sends once a second to all peers
transactions' hashes, which have not been sent. Peers
then request missing transactions from the node.

### Context of Change

This is a new feature. The architecture is based on a data 
modeling analysis of the crawled XRPL network. The analysis 
shows a densely connected core with the rest of the nodes
having multiple connections to the nodes in the core. Random
selection of peers and periodic pull of missing transactions
reduces redundancy in the transaction's propagation.

### Type of Change

- [x ] New feature (non-breaking change which adds functionality)

### CODE REVIEW

File: src/ripple/overlay/impl/OverlayImpl.cpp [hunk 1]
```diff
@@ -1298,6 +1324,65 @@ OverlayImpl::relay(
...
+        // select a fraction of active peers for relaying
+        auto toSelect = static_cast<uint32_t>(
+            app_.config().TX_RELAY_TO_PEERS * peers.size() / 100);
+
+        txMetrics_.addMetrics(toSelect, toSkip.size(), notEnabled);
+
+        // exclude from peers, which should be randomly selected for relaying,
+        // peers in toSkip and peers with not enabled tx reduce-relay feature
+        auto const exclude = toSkip.size() + notEnabled;
+        toSelect = (toSelect > exclude) ? (toSelect - exclude) : 0;
```
Comment (pwang200) [inline src/ripple/overlay/impl/OverlayImpl.cpp:1351]: Could you please explain the calculation of `toSelect`? I think the part that confused me is the definition of `TX_RELAY_TO_PEERS`, i.e. "Percentage of peers to select from total peers" as in config.h. My understanding is that if a node has 100 peers and `TX_RELAY_TO_PEERS`=50, then the node wants to make sure 50 of its peers have the Tx. Either they have the Tx already or the node will forward the Tx to them. 

Now consider a case that `toSkip.size()` = 40 and `notEnabled` = 0. Then `exclude` = 40, `peers.size()` = 60, and `TX_RELAY_TO_PEERS * peers.size() / 100` = 50*60/100 = 30. And 30 < 40, so `toSelect` will be 0. So no forwarding now. Hope my calculation is correct.

**Review by pwang200 — COMMENTED** [review]
Overall looks good.

The code is difficult to unit test. I wonder what cases were tested in a network.

File: src/ripple/proto/ripple.proto [hunk 2]
```diff
@@ -348,3 +356,8 @@ message TMSquelch
     optional uint32 squelchDuration = 3; // squelch duration in milliseconds
 }
 
+message TMHaveTransactions
+{
+    repeated bytes hash = 1;
```
[the remark this answers is not in the document]
Comment (pwang200) [reply src/ripple/proto/ripple.proto]: yeah.

**Review by nbougalis — CHANGES_REQUESTED** [review]

**Review by pwang200 — COMMENTED** [review]
I added several more comments and a placeholder for resuming the review once the handshake PR is merged.

File: src/ripple/overlay/impl/PeerImp.cpp [hunk 3]
```diff
@@ -1439,6 +1526,11 @@ PeerImp::onMessage(std::shared_ptr<protocol::TMTransaction> const& m)
                 JLOG(p_journal_.debug()) << "Ignoring known bad tx " << txID;
             }
```
[the remark this answers is not in the document]
Comment (gregtatcam) [reply src/ripple/overlay/impl/PeerImp.cpp:1527]: will update.

File: src/ripple/overlay/impl/OverlayImpl.cpp [hunk 4]
```diff
@@ -1298,6 +1331,68 @@ OverlayImpl::relay(
...
+    }
+
+    std::size_t disabled = 0;
+    std::size_t disabledInSkip = 0;
+    // active peers excluding peers in toSkip
+    auto peers = getActivePeers(toSkip, disabled, disabledInSkip);
+
+    // select a fraction of all active peers with the feature enabled
+    auto toRelay = static_cast<std::size_t>(
+        app_.config().TX_RELAY_PERCENTAGE * (active - disabled) / 100);
```
[the remark this answers is not in the document]
Comment (gregtatcam) [reply src/ripple/overlay/impl/OverlayImpl.cpp:1363]: actually more like this: `toRelay = min + disabled + pct * (active - min - disabled) ` assuming `active > min + disabled`. we don't want to include `disabled` in the random sample to relay.

[the remark this answers is not in the document]
Comment (nbougalis) [reply src/ripple/overlay/impl/OverlayImpl.cpp:1363]: Can we simply have a "target" value and simply reduce gradually? Something like: `(current - target) / 4`?

File: src/ripple/overlay/impl/PeerImp.h [hunk 5]
```diff
@@ -199,6 +200,16 @@ class PeerImp : public Peer,
 
     Compressed compressionEnabled_ = Compressed::Off;
 
+    // Queue of transactions' hashes that have not been
+    // relayed. The hashes are sent once a second to a peer
+    // and the peer requests missing transactions from the node.
+    hash_set<uint256> txQueue_;
```
[the remark this answers is not in the document]
Comment (nbougalis) [reply src/ripple/overlay/impl/PeerImp.h:206]: :+1:

**Review by nbougalis — APPROVED** [review]

**Comment by nbougalis** [comment]
@gregtatcam can you rebase this on the latest tip of **develop** please?

**Review by pwang200 — APPROVED** [review]
:+1: Looks good!

**Review by pwang200 — APPROVED** [review]
:+1: LGTM

**Comment by troylougheed** [comment]
Unsubscribe￼

**Pull request closed without merging**